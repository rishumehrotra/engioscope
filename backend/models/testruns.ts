import { head, last, range } from 'rambda';
import { z } from 'zod';
import { oneDayInMs, oneWeekInMs } from '../../shared/utils.js';
import { collectionAndProjectInputs, dateRangeInputs, inDateRange } from './helpers.js';
import { BuildDefinitionModel } from './mongoose-models/BuildDefinitionModel.js';
import { RepositoryModel } from './mongoose-models/RepositoryModel.js';

export type TestStatDetails = {
  state: string;
  outcome: string;
  count: number;
};

export type BuildPipelineTests = {
  _id: number;
  collectionName: string;
  project: string;
  repositoryId: string;
  repositoryName: string;
  defaultBranch: string;
  buildId: number;
  sourceBranch: string;
  definitionId: number;
  definitionName: string;
  definitionUrl: string;
  testDuration: number;
  totalTests: number;
  testStats: TestStatDetails[];
  testsName: string;
};
export const TestRunsForRepositoryInputParser = z.object({
  ...collectionAndProjectInputs,
  repositoryId: z.string(),
  ...dateRangeInputs,
});

type BuildDef = { id: number; name: string; url: string };

type TestsForWeek = {
  weekIndex: number;
  totalTests: number;
  startedDate: Date | null;
  completedDate: Date | null;
  passedTests: number;
  hasTests: boolean;
};
type TestsForDef = {
  definitionId: number;
  tests: TestsForWeek[];
  latest?: TestsForWeek;
};
type BuildDefWithTests = BuildDef & Partial<TestsForDef>;

const makeContinuous = async (
  tests: TestsForWeek[] | undefined,
  startDate: Date,
  endDate: Date,
  getOneOlderTestRun: () => Promise<TestsForWeek | null>
) => {
  const totalDays = (endDate.getTime() - startDate.getTime()) / oneDayInMs;
  const totalIntervals = Math.floor(totalDays / 7 + (totalDays % 7 === 0 ? 0 : 1));

  if (!tests) {
    const olderTest = await getOneOlderTestRun();
    if (!olderTest) return null;

    return range(0, totalIntervals).map(weekIndex => {
      return { ...olderTest, weekIndex };
    });
  }

  return range(0, totalIntervals)
    .reduce<Promise<TestsForWeek[]>>(async (acc, weekIndex, index) => {
      const matchingTest = tests.find(t => t.weekIndex === weekIndex);

      if (matchingTest) return [...(await acc), matchingTest];

      if (index === 0) {
        const olderTest = await getOneOlderTestRun();

        if (!olderTest) {
          return [
            {
              weekIndex,
              totalTests: 0,
              passedTests: 0,
              startedDate: null,
              completedDate: null,
              hasTests: false,
            },
          ];
        }

        return [{ ...olderTest, weekIndex }];
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const lastItem = last(await acc)!;
      return [...(await acc), { ...lastItem, weekIndex }];
    }, Promise.resolve([]))
    .then(list => list.slice(totalIntervals - Math.floor(totalDays / 7)));
};

export const getOldTestRunsForDefinition = async (
  collectionName: string,
  project: string,
  startDate: Date,
  repositoryId: string,
  definitionId: number
) => {
  const result = await RepositoryModel.aggregate<TestsForWeek>([
    {
      $match: {
        'collectionName': collectionName,
        'project.name': project,
        'id': repositoryId,
      },
    },
    {
      $project: {
        _id: 0,
        collectionName: '$collectionName',
        project: '$project.name',
        repositoryId: '$id',
        repositoryName: '$name',
        defaultBranch: '$defaultBranch',
      },
    },
    {
      $lookup: {
        from: 'builds',
        let: {
          collectionName: '$collectionName',
          project: '$project',
          repositoryId: '$repositoryId',
          defaultBranch: '$defaultBranch',
        },
        pipeline: [
          {
            $match: {
              $expr: {
                $and: [
                  { $eq: ['$collectionName', '$$collectionName'] },
                  { $eq: ['$project', '$$project'] },
                  { $eq: ['$repository.id', '$$repositoryId'] },
                  { $eq: ['$sourceBranch', '$$defaultBranch'] },
                  // Different from the original query
                  { $eq: ['$definition.id', definitionId] },
                  { $lt: ['$finishTime', new Date(startDate)] },
                ],
              },
            },
          },
          { $sort: { finishTime: -1 } },
          {
            $project: {
              _id: 0,
              buildId: '$id',
              sourceBranch: '$sourceBranch',
              definitionId: '$definition.id',
              definitionName: '$definition.name',
              result: '$result',
              finishTime: '$finishTime',
            },
          },
        ],
        as: 'build',
      },
    },
    { $unwind: { path: '$build' } },
    {
      $group: {
        _id: {
          definitionId: '$build.definitionId',
          weekIndex: {
            $trunc: {
              $divide: [
                { $subtract: ['$build.finishTime', new Date(startDate)] },
                oneWeekInMs,
              ],
            },
          },
        },
        collectionName: { $first: '$collectionName' },
        project: { $first: '$project' },
        repositoryId: { $first: '$repositoryId' },
        repositoryName: { $first: '$repositoryName' },
        build: { $first: '$build' },
      },
    },
    {
      $lookup: {
        from: 'testruns',
        let: {
          collectionName: '$collectionName',
          project: '$project',
          buildId: '$build.buildId',
        },
        pipeline: [
          {
            $match: {
              release: { $exists: false },
              $expr: {
                $and: [
                  { $eq: ['$collectionName', '$$collectionName'] },
                  { $eq: ['$project.name', '$$project'] },
                  { $eq: ['$buildConfiguration.id', '$$buildId'] },
                ],
              },
            },
          },
          {
            $addFields: {
              passed: {
                $filter: {
                  input: '$runStatistics',
                  as: 'stats',
                  cond: { $eq: ['$$stats.outcome', 'Passed'] },
                },
              },
            },
          },
          {
            $addFields: {
              passedCount: { $sum: '$passed.count' },
            },
          },
        ],
        as: 'tests',
      },
    },
    {
      $project: {
        _id: 0,
        definitionId: '$_id.definitionId',
        weekIndex: '$_id.weekIndex',
        totalTests: { $sum: '$tests.totalTests' },
        startedDate: { $min: '$tests.startedDate' },
        completedDate: { $max: '$tests.completedDate' },
        passedTests: { $sum: '$tests.passedCount' },
        buildId: '$build.buildId',
        testIds: {
          $reduce: {
            input: '$tests.id',
            initialValue: [],
            in: { $setUnion: ['$$value', ['$$this']] },
          },
        },
      },
    },
    {
      $addFields: {
        hasTests: { $gt: [{ $size: '$testIds' }, 0] },
        foundTests: {
          $and: [{ $gt: [{ $size: '$testIds' }, 0] }, { $ne: ['$completedDate', null] }],
        },
      },
    },
    { $sort: { weekIndex: -1 } },
    {
      $match: {
        totalTests: { $gt: 0 },
      },
    },
    {
      $limit: 1,
    },
    {
      $project: {
        definitionId: 0,
        testIds: 0,
        buildId: 0,
      },
    },
  ]);

  return head(result) || null;
};

export const getTestRunsForRepo = async (
  collectionName: string,
  project: string,
  startDate: Date,
  endDate: Date,
  repositoryId: string
) => {
  const [definitionList, definitionTestRuns] = await Promise.all([
    BuildDefinitionModel.find(
      {
        collectionName,
        project,
        repositoryId,
      },
      {
        _id: 0,
        id: 1,
        name: 1,
        url: 1,
      }
    ).lean(),
    RepositoryModel.aggregate<TestsForDef>([
      {
        $match: {
          'collectionName': collectionName,
          'project.name': project,
          'id': repositoryId,
        },
      },
      {
        $project: {
          _id: 0,
          collectionName: '$collectionName',
          project: '$project.name',
          repositoryId: '$id',
          repositoryName: '$name',
          defaultBranch: '$defaultBranch',
        },
      },
      {
        $lookup: {
          from: 'builds',
          let: {
            collectionName: '$collectionName',
            project: '$project',
            repositoryId: '$repositoryId',
            defaultBranch: '$defaultBranch',
          },
          pipeline: [
            {
              $match: {
                $expr: {
                  $and: [
                    { $eq: ['$collectionName', '$$collectionName'] },
                    { $eq: ['$project', '$$project'] },
                    { $eq: ['$repository.id', '$$repositoryId'] },
                    { $eq: ['$sourceBranch', '$$defaultBranch'] },
                  ],
                },
                finishTime: inDateRange(startDate, endDate),
              },
            },
            { $sort: { finishTime: -1 } },
            {
              $project: {
                _id: 0,
                buildId: '$id',
                sourceBranch: '$sourceBranch',
                definitionId: '$definition.id',
                definitionName: '$definition.name',
                result: '$result',
                finishTime: '$finishTime',
              },
            },
          ],
          as: 'build',
        },
      },
      { $unwind: { path: '$build' } },
      {
        $group: {
          _id: {
            definitionId: '$build.definitionId',
            weekIndex: {
              $trunc: {
                $divide: [
                  { $subtract: ['$build.finishTime', new Date(startDate)] },
                  oneWeekInMs,
                ],
              },
            },
          },
          collectionName: { $first: '$collectionName' },
          project: { $first: '$project' },
          repositoryId: { $first: '$repositoryId' },
          repositoryName: { $first: '$repositoryName' },
          build: { $first: '$build' },
        },
      },
      {
        $lookup: {
          from: 'testruns',
          let: {
            collectionName: '$collectionName',
            project: '$project',
            buildId: '$build.buildId',
          },
          pipeline: [
            {
              $match: {
                release: { $exists: false },
                $expr: {
                  $and: [
                    { $eq: ['$collectionName', '$$collectionName'] },
                    { $eq: ['$project.name', '$$project'] },
                    { $eq: ['$buildConfiguration.id', '$$buildId'] },
                  ],
                },
              },
            },
            {
              $addFields: {
                passed: {
                  $filter: {
                    input: '$runStatistics',
                    as: 'stats',
                    cond: { $eq: ['$$stats.outcome', 'Passed'] },
                  },
                },
              },
            },
            {
              $addFields: {
                passedCount: { $sum: '$passed.count' },
              },
            },
          ],
          as: 'tests',
        },
      },
      {
        $project: {
          _id: 0,
          definitionId: '$_id.definitionId',
          weekIndex: '$_id.weekIndex',
          totalTests: { $sum: '$tests.totalTests' },
          startedDate: { $min: '$tests.startedDate' },
          completedDate: { $max: '$tests.completedDate' },
          passedTests: { $sum: '$tests.passedCount' },
          buildId: '$build.buildId',
          testIds: {
            $reduce: {
              input: '$tests.id',
              initialValue: [],
              in: { $setUnion: ['$$value', ['$$this']] },
            },
          },
        },
      },
      {
        $addFields: {
          hasTests: { $gt: [{ $size: '$testIds' }, 0] },
          foundTests: {
            $and: [
              { $gt: [{ $size: '$testIds' }, 0] },
              { $ne: ['$completedDate', null] },
            ],
          },
        },
      },
      { $sort: { weekIndex: -1 } },
      {
        $group: {
          _id: '$definitionId',
          definitionId: { $first: '$definitionId' },
          tests: { $push: '$$ROOT' },
        },
      },
      {
        $project: {
          '_id': 0,
          'tests.definitionId': 0,
          'tests.testIds': 0,
          'tests.buildId': 0,
        },
      },
    ]),
  ]);

  // Mapping the build definitions/pipelines with no testruns
  const buildDefsWithTests: BuildDefWithTests[] = (definitionList as BuildDef[]).map(
    definition => {
      const tests = definitionTestRuns.find(def => def.definitionId === definition.id);
      return { ...definition, ...tests } || definition;
    }
  );
  return buildDefsWithTests;
};

export const getTestRunsForRepository = async ({
  collectionName,
  project,
  repositoryId,
  startDate,
  endDate,
}: z.infer<typeof TestRunsForRepositoryInputParser>) => {
  const testRunsForRepo = await getTestRunsForRepo(
    collectionName,
    project,
    startDate,
    endDate,
    repositoryId
  );

  console.log(
    'testRunsForRepo',
    testRunsForRepo.map(t => t.tests)
  );

  const getOneOlderTestRunForDef = (defId: number) => () => {
    return getOldTestRunsForDefinition(
      collectionName,
      project,
      startDate,
      repositoryId,
      defId
    );
  };

  const definitionTests = Promise.all(
    testRunsForRepo.map(async def => {
      const tests = await makeContinuous(
        def.tests,
        startDate,
        endDate,
        getOneOlderTestRunForDef(def.id)
      );

      return {
        ...def,
        tests,
        latest: tests ? last(tests) : null,
      };
    })
  );
  return definitionTests;
};
