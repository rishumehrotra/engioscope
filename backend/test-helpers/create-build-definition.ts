import { oneMinuteInMs, oneSecondInMs } from '../../shared/utils.js';
import { BuildDefinitionModel } from '../models/mongoose-models/BuildDefinitionModel.js';

export const createLatestBuildForDefinition = (
  buildId: number,
  finishTime: Date,
  status = 'completed',
  result = 'succeeded',
  buildDuration = 2 * oneMinuteInMs
) => {
  const startTime = new Date(finishTime);
  startTime.setTime(finishTime.getTime() - buildDuration);

  const queueTime = new Date(finishTime);
  startTime.setTime(finishTime.getTime() - buildDuration - 30 * oneSecondInMs);

  return {
    id: buildId,
    status,
    result,
    queueTime,
    startTime,
    finishTime,
  };
};

export const createBuildDefinition = (
  collectionName: string,
  project: string,
  buildDefinitionId: number,
  repositoryId: string,
  isYaml?: boolean,
  latestBuild = createLatestBuildForDefinition(123, new Date('2022-03-25')),
  latestCompletedBuild = createLatestBuildForDefinition(123, new Date('2022-03-25'))
) => {
  return BuildDefinitionModel.insertMany([
    {
      collectionName,
      project,
      id: buildDefinitionId,
      name: 'build def 1',
      url: 'http://foo.bar',
      projectId: 'project-1',
      repositoryId,
      createdDate: new Date('2020-01-01'),
      queueStatus: 'enabled',
      revision: '123',
      type: 'build',
      uri: 'vstfs://foo.bar',
      latestBuild,
      latestCompletedBuild,
      process: isYaml ? { processType: 2, yamlFilename: 'foo.yaml' } : { processType: 1 },
    },
  ]);
};
