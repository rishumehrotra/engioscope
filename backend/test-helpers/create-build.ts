import { oneMinuteInMs, oneSecondInMs } from '../../shared/utils.js';
import { BuildModel } from '../models/mongoose-models/BuildModel.js';

export const createBuild = (
  collectionName: string,
  project: string,
  repoId: string,
  buildId: number,
  buildDefinitionId: number,
  finishTime = new Date('2022-03-25'),
  buildDuration = 2 * oneMinuteInMs
) => {
  const startTime = new Date(finishTime);
  startTime.setTime(finishTime.getTime() - buildDuration);

  const queueTime = new Date(finishTime);
  startTime.setTime(finishTime.getTime() - buildDuration - 30 * oneSecondInMs);

  return BuildModel.insertMany([
    {
      id: buildId,
      collectionName,
      project,
      repository: {
        id: repoId,
        name: 'repo-1',
      },
      buildNumber: `build-${buildId}`,
      buildNumberRevision: 1,
      definition: { id: buildDefinitionId, name: 'build def 1', url: 'http://foo.bar' },
      finishTime,
      queueTime,
      reason: 'manual',
      result: 'success',
      sourceBranch: 'refs/heads/master',
      sourceVersion: 'some-string',
      startTime,
      status: 'completed',
      uri: 'vstfs://Build/Build/123',
      url: 'http://example.com/foo/bar',
    },
  ]);
};
