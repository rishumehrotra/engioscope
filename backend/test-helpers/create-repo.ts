import { RepositoryModel } from '../models/mongoose-models/RepositoryModel.js';
import { generateRandomTestId } from './utils.js';

export const createRepo = (
  collectionName: string,
  project: string,
  repoId: string,
  defaultBranch = 'refs/heads/master'
) => {
  return RepositoryModel.insertMany([
    {
      id: repoId,
      name: 'repo-1',
      collectionName,
      project: {
        id: `project-${generateRandomTestId()}`,
        lastUpdatedTime: new Date('2022-01-01'),
        name: project,
        state: 'wellFormed',
        visibility: 'public',
      },
      remoteUrl: 'http://example.com/',
      size: 123,
      sshUrl: 'ssh://example.com',
      url: 'http://example.com',
      webUrl: 'http://example.com',
      defaultBranch,
    },
  ]);
};
