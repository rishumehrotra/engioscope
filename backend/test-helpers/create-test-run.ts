import { TestRunModel } from '../models/mongoose-models/TestRunModel.js';
import { generateRandomTestId } from './utils.js';

export const createTestRun = (
  collectionName: string,
  project: string,
  buildId: number,
  buildDefinitionId: number,
  totalTests = 10
) => {
  return TestRunModel.insertMany([
    {
      collectionName,
      id: generateRandomTestId(),
      project: {
        id: 'project-1',
        name: project,
      },
      buildConfiguration: {
        id: buildId,
        number: 'master_20230220.1',
        platform: '',
        buildDefinitionId,
        project: {
          name: 'project-1',
        },
      },
      completedDate: new Date('2022-03-25'),
      name: 'JUnit_TestResults_983411',
      revision: 0,
      runStatistics: [
        {
          state: 'Unspecified',
          outcome: 'Passed',
          count: totalTests,
        },
      ],
      startedDate: new Date('2022-03-25'),
      state: 'Completed',
      totalTests,
      url: 'http://example.com/foo',
      webAccessUrl: 'https://example.com/foo',
    },
  ]);
};

export const getTestruns = (collectionName: string, project: string, buildId: number) =>
  TestRunModel.find({
    collectionName,
    'project.name': project,
    'buildConfiguration.id': buildId,
  });
