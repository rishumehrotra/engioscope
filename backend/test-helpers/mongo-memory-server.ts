/* eslint-disable import/no-extraneous-dependencies */
import mongoose from 'mongoose';
import { afterAll, beforeAll, beforeEach } from 'vitest';
import { setup, teardown } from 'vitest-mongodb';

// eslint-disable-next-line no-underscore-dangle
declare let __MONGO_URI__: string;

export const needsDB = () => {
  beforeAll(async () => {
    await setup();

    await mongoose.connect(__MONGO_URI__);
  });

  afterAll(async () => {
    await teardown();
    await mongoose.connection.close();
  });

  beforeEach(async () => {
    await mongoose.connection.db.dropDatabase();
  });
};
