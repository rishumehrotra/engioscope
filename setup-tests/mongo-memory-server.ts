/* eslint-disable import/no-extraneous-dependencies */
import mongoose from 'mongoose';
import { beforeEach, afterEach } from 'vitest';
import { setup, teardown } from 'vitest-mongodb';

beforeEach(async () => {
  await setup();
  // eslint-disable-next-line no-underscore-dangle
  await mongoose.connect(globalThis.__MONGO_URI__);
});

afterEach(async () => {
  await teardown();
  await mongoose.connection.close();
});
