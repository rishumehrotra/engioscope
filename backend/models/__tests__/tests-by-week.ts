import { it, expect } from 'vitest';
import { createBuildDefinition } from '../../test-helpers/create-build-definition.js';
import { createBuild } from '../../test-helpers/create-build.js';
import { createRepo } from '../../test-helpers/create-repo.js';
import { createTestRun, getTestruns } from '../../test-helpers/create-test-run.js';
import { needsDB } from '../../test-helpers/mongo-memory-server.js';
import { getTestsAndCoverageByWeek } from '../testruns.js';

needsDB();

it('Given buildId should give a testrun for it', async () => {
  // Arrange
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 123, 12_345, new Date('2022-03-25'));
  await createTestRun('foo', 'bar', 123, 12_345);
  // Act
  const tests = await getTestruns('foo', 'bar', 123);
  // Assert
  expect(tests.length).toBe(1);
});

it('should give right no of weeks for query period', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-03-25'));
  await createTestRun('foo', 'bar', 122, 12_345);

  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );

  expect(testsByWeek.length).toBe(12);
});

it('should not give test counts when there are no testruns', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-03-25'));

  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );

  expect(testsByWeek.length).toBe(0);
});

it('should give test counts when there are testruns', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-01-02'));
  await createTestRun('foo', 'bar', 122, 12_345);
  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );
  expect(testsByWeek.length).toBe(12);
  expect(testsByWeek.every(t => t.hasTests && t.passedTests === 10)).toBe(true);
  expect(testsByWeek.every(t => t.hasTests && t.totalTests === 10)).toBe(true);
});

it('should handle missing test run counts with no past data', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-02-15'));

  await createTestRun('foo', 'bar', 122, 12_345);

  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );

  expect(testsByWeek.length).toBe(12);
  expect(testsByWeek.slice(0, 6).every(t => !t.hasTests)).toBe(true);
  expect(testsByWeek.slice(6).every(t => t.hasTests && t.passedTests === 10)).toBe(true);
  expect(testsByWeek.slice(6).every(t => t.hasTests && t.totalTests === 10)).toBe(true);
});

it('should handle missing test run counts with past data', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 121, 12_345, new Date('2021-12-15'));
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-02-30'));

  await createTestRun('foo', 'bar', 122, 12_345);

  await createTestRun('foo', 'bar', 121, 12_345, 15);

  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );

  expect(testsByWeek.length).toBe(12);
  expect(testsByWeek.slice(0, 8).every(t => t.hasTests && t.passedTests === 15)).toBe(
    true
  );
  expect(testsByWeek.slice(0, 8).every(t => t.hasTests && t.totalTests === 15)).toBe(
    true
  );

  expect(testsByWeek.slice(8).every(t => t.hasTests && t.passedTests === 10)).toBe(true);
  expect(testsByWeek.slice(8).every(t => t.hasTests && t.totalTests === 10)).toBe(true);
});

it('should handle missing test run', async () => {
  await createRepo('foo', 'bar', 'repo-1');
  await createBuildDefinition('foo', 'bar', 12_345, 'repo-1');
  await createBuild('foo', 'bar', 'repo-1', 122, 12_345, new Date('2022-02-30'));
  await createBuild('foo', 'bar', 'repo-1', 123, 12_345, new Date('2022-03-20'));
  await createTestRun('foo', 'bar', 122, 12_345);

  const { testsByWeek } = await getTestsAndCoverageByWeek(
    ['foo', 'bar', new Date('2022-01-01'), new Date('2022-03-26')],
    ['repo-1']
  );

  expect(testsByWeek).toMatchSnapshot();
});
