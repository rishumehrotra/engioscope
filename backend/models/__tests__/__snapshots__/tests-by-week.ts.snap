// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should handle missing test run 1`] = `
[
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 1,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 2,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 3,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 4,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 5,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 6,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 7,
  },
  {
    "passedTests": 10,
    "totalTests": 10,
    "weekIndex": 8,
  },
  {
    "passedTests": 10,
    "totalTests": 10,
    "weekIndex": 9,
  },
  {
    "passedTests": 10,
    "totalTests": 10,
    "weekIndex": 10,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 11,
  },
  {
    "passedTests": 0,
    "totalTests": 0,
    "weekIndex": 12,
  },
]
`;
